### ISSUE TITLE
Массоввое появление подчиненных строк с нулевым SYSN в Журналах техн.операций

### ISSUE THREAD

**Issue by BMKasu1** [savushkin-r-d/INER#22]
Чаще всего это «явление» происходит по ночам, хотя сегодня вот утром тоже такое случилось.
Проблема следующая: время от времени поступают звонки от лаборатории, мол не можем работать с Журналами, т.к.
в базе Журналы технологических операций невозможно добавить строку, выдается «Ошибка при добавлении записи» и внизу не генерируется список показателей к вводимой операции.

пока дежурный не подключится и все эти подчиненные строки не удалит, никто работать не может.
Явление это началось после перевода Журналов на SQL (перешли 9 декабря 2020года).

Никак не могли понять, откуда эти нулевые SYSN вообще могут взяться, пользователи внятно не отвечали, после каких действий Журналы «накрывало».
Катя перешерстила обработчики, явных обнулений SYSN не нашли.
Действовали симптоматично (удаляли строки BD_PLGTH с нулевым SYSN) и периодически пытались имитировать подобную ситуацию, но как-то не могли додуматься что ж они такое делали.

И НАКОНЕЦ-ТО СЕГОДНЯ на меня снизошло озарение))))) и получилось-таки найти эти пользовательские телодвижения, которые приводили в массовому появлению строк в BD_PLGTH c нулевыми SYSN.

Матсклад (j:\tmc\tmcQQ+.exe) – попробовала так же на exe от 17-02-2021
Документы – Производство – Журнал технологических операций (БАЗА РЕАЛЬНАЯ, АКТИВНО ИСПОЛЬЗУЕТСЯ)
![1](https://user-images.githubusercontent.com/82371353/115025832-7c3eb980-9eca-11eb-93dd-4dab53b3c37e.png)

В дереве рабочих центров (слева) выбрать «20 ОРША» (в Бресте не используется, годится для тестов)
![2](https://user-images.githubusercontent.com/82371353/115025898-8e205c80-9eca-11eb-9c2a-a3588e0b9055.png)

После этого нажимаем Insert. Появляется строка-заголовок. Автоматически заполнено Nп/п очередным SYSN, в поле РЦ проставился 200000000000 ОРША.
В поле Операция вносим цифру 1 (операция приемка сырья) и нажимаем Enter.
Появляется список строк подчиненной таблице. В заголовке проставляем Дату начала  16.04.2021 и время начала 9.00 (к примеру сегодняшний день).
![3](https://user-images.githubusercontent.com/82371353/115025960-9d9fa580-9eca-11eb-9c50-29e6a9e5c510.png)

Переходим на подчиненную таблицу.

(Пользователь заполняет как правило колонку данные, если нужен текст – то примечание) – можем вовсе ничего в строках не заполнять, для нашего теста не критично. 

Переходим на последнюю строку в подчиненной базе и нажимаем вниз. Появляется новая строка, без показателя (можем туда и внести что-то, к примеру P_VP это не повлияет никак на ход теста).
В колонке «Данные» вносим 777 (или что-то другое, к примеру 0) и нажимаем Enter(произойдет переход на след.поле «Примечание») – именно так, строка в подчиненной базе должна оставаться со *.
![4](https://user-images.githubusercontent.com/82371353/115026019-aa23fe00-9eca-11eb-8ddb-701ce1edb47a.png)

Теперь нажимаем F2. (можно находиться в подчиненной, это не важно – происходит сохранение заголовка, на него навешен просчет операции).
На первый взгляд ничего страшного не происходит, запись сохраняется. Но вот эта самая последняя строка «исчезает» из списка подчиненных:

**(!!!)** И теперь, если мы попытаемся внести новую строку-заголовок, она появится, но выдаст ошибку и в списке подчиненных окажется видна наша строчка (при ближнем рассмотрении – у нее SYSN=0)
![6](https://user-images.githubusercontent.com/82371353/115028074-21f32800-9ecd-11eb-9b73-8e64c3173815.png)

Жму ОК. Удаляю подчиненную строку с «777» в Данных. Она имеет SYSN=0 (можно в подчиненной базе визуализировать столбец Nп/п, чтобы явно было видно)

Добавляю строку-заголовок через Insert.
Вношу в РЦ 200000000000 (ОРША), в Операцию – 1. 
![8](https://user-images.githubusercontent.com/82371353/115028100-2b7c9000-9ecd-11eb-808e-d56dd48dde7f.png)

После ввода операции в подчиненной таблице должен подцепиться список показателей (как на скриншоте выше, 16 строк). Но наши строки в подч.базе увы не появились, поэтому назовем их «потеряшки».
![9](https://user-images.githubusercontent.com/82371353/115028847-01779d80-9ece-11eb-95c7-6e7a45674906.png)

Удаляем строку-заголовок.
![10](https://user-images.githubusercontent.com/82371353/115028901-0dfbf600-9ece-11eb-9e5f-1e051fd0883b.png)

Опять пытаюсь добавить новую строку-заголовок – Ошибка при добавлении записи (подцепились строки с нулевым SYSN, которые мы назвали выше «потеряшками»)
![11](https://user-images.githubusercontent.com/82371353/115028933-194f2180-9ece-11eb-9871-0850ad221085.png)

Посмотрим обработчик на изменение KOP (где рождается перечень подчиненных строк) – там четко проставляется SYSN из строки-заголовка.
![13](https://user-images.githubusercontent.com/82371353/115029126-55828200-9ece-11eb-816e-82ce9ec988dd.png)

Выше коду описано, что в l_sysn:
![14](https://user-images.githubusercontent.com/82371353/115029131-574c4580-9ece-11eb-8653-0b658dafe58b.png)

Останавливалась на debug в момент добавления подчиненных строк, которые становились «потеряшками» (см. где пометка выше (!!!)), включала BD_PLGTH->Dbobj.EnableControls и смотрела, SYSN действительно перед DbPost заполнен обработчиком верно, SYSN:=8021777, но после выполнения DbPost он каким-то чудом превращается в SYSN=0.

Похоже, это связано с имитацией транзакций для таблиц sql.

При вводе строки-заголовка (см. где пометка выше (!!!)) ведь выдавало ошибку при добавлении записи, но все же строка была добавлена и в BD_ZLGTH->SYSN попал очередной номер 8021777.
Похоже ваш внутренний механизм из-за этой ошибки таки не инициализировал ключ SYSN в 8021777, а искренне полагал, что ключ SYSN=0 и при сохранении вот брал себе и «поверх моего 8021777» уверенно писал свой «неинициализированный ключ SYSN=0».

**Итого два вопроса:** 
**1)** при сохранении строки-заголовка, должна сохраняться все подчиненные строки в текущий ключ SYSN (а у нас подчиненная строка, которая остается в состоянии *новой строки, так сохраняется в базу с неинициализированным ключом SYSN=0)
**2)** при возникновении ошибки при добавлении строки-заголовка (что вызвано наличием подчиненных строк с SYSN=0) похоже, не инициализируется текущий ключ для имитации транзакций (моё предположение). А это в дальнейшем приводит к массовому сохранению строк с пустым SYSN, и  база по сути становится «парализованной» для работы.

**Comment by kostomarovvv** [comment]
j:\TMCN

**Comment by kostomarovvv** [comment]
Причина ошибки в том, что при вставке срабатывает условие (активен DisableControls):
`if (not TxTable(Dataset).Mastersource.Dataset.ControlsDisabled)`
в методе Tdocbd.Detailinsert, и InitRecord не вызывается, соответственно не заполняется поле SYSN.

![image](https://user-images.githubusercontent.com/17332806/116214465-66918580-a74f-11eb-9a44-40d81370b244.png)

**Comment by BMKasu1** [comment]
Да, коммент строки BD_PLGTH.disablecontrols в обработчике BeforePost для BD_ZLGTH снимает ошибку. 
![2q22](https://user-images.githubusercontent.com/82371353/126990992-e766a14c-6813-4224-aa82-28a408a391cc.png)

Но по смыслу disablecontrols должен быть.
Перемещение по строчкам в момент отработки BeforePost может исказить расчет показателей. Т.к. при сохранении заголовка (BD_ZLGTH) происходит "пробежка" по внесенным показателям (BD_PLGTH) + расчет и добавление новых показателей (например, среднее и т.п. по формулам).

Пока не понимаю, почему в exe условие именно на disablecontrols..
